homeassistant/components/integration/sensor.py

```diff
@@ -89,6 +90,72 @@
 )
 
 
+class _IntegrationMethod(ABC):
+    @staticmethod
+    def from_name(method_name: str) -> _IntegrationMethod:
+        if method_name == METHOD_TRAPEZOIDAL:
+            return _Trapezoidal()
+        if method_name == METHOD_RIGHT:
+            return _Right()
+        return _Left()
+
+    @abstractmethod
+    def validate_states(self, left: State | None, right: State | None) -> bool:
+        """Check state requirements for integration."""
+
+    @abstractmethod
+    def calculate_area_with_two_states(
+        self, elapsed_time: float, left: State, right: State
+    ) -> Decimal:
+        """Calculate area given two states."""
+
+    def calculate_area_with_one_state(
+        self, elapsed_time: float, constant_state: State
+    ) -> Decimal:
+        return Decimal(constant_state.state) * Decimal(elapsed_time)
+
+
+class _Trapezoidal(_IntegrationMethod):
+    def calculate_area_with_two_states(
+        self, elapsed_time: float, left: State, right: State
+    ) -> Decimal:
+        return Decimal(elapsed_time) * (Decimal(left.state) + Decimal(right.state)) / 2
+
+    def validate_states(self, left: State | None, right: State | None) -> bool:
+        return _is_numeric_state(left) and _is_numeric_state(right)
+
+
+class _Left(_IntegrationMethod):
+    def calculate_area_with_two_states(
+        self, elapsed_time: float, left: State, right: State
+    ) -> Decimal:
+        return self.calculate_area_with_one_state(elapsed_time, left)
+
+    def validate_states(self, left: State | None, right: State | None) -> bool:
+        return _is_numeric_state(left)
+
+
+class _Right(_IntegrationMethod):
+    def calculate_area_with_two_states(
+        self, elapsed_time: float, left: State, right: State
+    ) -> Decimal:
+        return self.calculate_area_with_one_state(elapsed_time, right)
+
+    def validate_states(self, left: State | None, right: State | None) -> bool:
+        return _is_numeric_state(right)
+
+
+def _is_numeric_state(state: State | None) -> bool:
+    if state is None:
+        return False
```

Comment:
Can we cover this case in tests?

Reply:
This actually revealed that the None type was unnecessary, so removed the None type from the method argument and removed this if-statement.